File: src/main/java/etlmail/engine/css/CssInliner.java
package etlmail.engine.css;

import static java.util.Collections.reverseOrder;
import static java.util.Collections.sort;

import java.util.*;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.stereotype.Component;

@Component
public class CssInliner {
    public void inlineStyles(Document doc) {
	for (final Element style : doc.select("style")) {
	    final String styleRules = style.getAllElements().get(0).data().replaceAll("\n", "").trim();
	    final StringTokenizer st = new StringTokenizer(styleRules, "{}");
	    final List<CssRule> cssRules = new ArrayList<CssRule>();
	    while (st.countTokens() > 1) {
		final String selector = st.nextToken();
		final String properties = st.nextToken();
		for (final String simpleSelector : selector.split(",")) {
		    cssRules.add(new CssRule(simpleSelector, properties));
		}
	    }
	    sort(cssRules, reverseOrder());
	    for (final CssRule rule : cssRules) {
		rule.prependProperties(doc);
	    }
	    style.remove();
	}
    }
}

File: src/main/scala/etlmail/engine/css/CssInliner.scala
package etlmail.engine.css

import java.util.Collections.reverseOrder
import java.util.Collections.sort

import java.util._

import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.springframework.stereotype.Component

import scala.collection.JavaConversions._

@Component
class CssInliner {
  def inlineStyles(doc: Document) {
    for (style <- doc.select("style")) {
      val styleRules = style.getAllElements().get(0).data().replaceAll("\n", "").trim()
      val st = new StringTokenizer(styleRules, "{}")
      val cssRules = new ArrayList[CssRule]()
      while (st.countTokens() > 1) {
        val selector = st.nextToken()
        val properties = st.nextToken()
        for (simpleSelector <- selector.split(",")) {
          cssRules.add(new CssRule(simpleSelector, properties))
        }
      }
      sort(cssRules, reverseOrder[CssRule]())
      for (rule <- cssRules) {
        rule.prependProperties(doc)
      }
      style.remove()
    }
  }
}
